### PR TITLE
#1476 - Display correct user & date of attachment of a file to a ticket

### LINKED ISSUE #1476: Wrong actor in ticket history
**Problem**
When i add an existing document to an existing ticket, under "Processing ticket > Actions historical" tab, the actor and datetime associated with the document on this ticket is not me but the creator and creation datetime of document.

**Configuration**
`
GLPI 9.1.1 ( => D:\APPLICATIFS\GLPI\WEBTEST)

Server
 
Operating system: Windows NT WENDY 6.1 build 7601 (Windows Server 2008 R2 Enterprise Edition Service Pack 1) i586
PHP 5.6.6 apache2handler (Core, PDO, Phar, Reflection, SPL, SimpleXML, Zend OPcache, apache2handler, bcmath, bz2, calendar,
	com_dotnet, ctype, curl, date, dom, ereg, fileinfo, filter, ftp, gd, hash, iconv, imap, json, ldap, libxml, mbstring, mcrypt,
	mhash, mysql, mysqli, mysqlnd, odbc, openssl, pcre, pdo_mysql, session, soap, sqlite3, standard, tokenizer, wddx, xml,
	xmlreader, xmlrpc, xmlwriter, xsl, zip, zlib)
Setup: max_execution_time="300" memory_limit="128M" post_max_size="8M" safe_mode="" session.save_handler="files"
	upload_max_filesize="2M" 
Software: Apache/2.4.12 (Win32) mod_authn_ntlm/1.0.0 PHP/5.6.6 ()
	Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/46.0.2490.80 Safari/537.36
Server Software: MySQL Community Server (GPL)
	Server Version: 5.5.19
	Server SQL Mode: 
	Parameters: root@127.0.0.1:3307/glpi_test
	Host info: 127.0.0.1 via TCP/IP
	
OKD:\APPLICATIFS\GLPI\WEBTEST/config : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_dumps : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_sessions : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_cron : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_graphs : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_lock : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_plugins : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_tmp : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_rss : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_uploads : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_pictures : OK
OKD:\APPLICATIFS\GLPI\WEBTEST/files/_log : OK

Libraries
 
htmLawed version 1.1.21 in (D:\APPLICATIFS\GLPI\WEBTEST\lib\htmlawed)
phpCas version 1.3.4 in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\jasig\phpcas\source)
PHPMailer version 5.2.16 in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\phpmailer\phpmailer)
Zend Framework in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\zendframework\zend-loader\src)
zetacomponents/graph in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\zetacomponents\graph\src)
SimplePie version 1.4.1 in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\simplepie\simplepie\library)
TCPDF version 6.2.12 in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\tecnickcom\tcpdf)
michelf/php-markdown in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\michelf\php-markdown\Michelf)
true/punycode in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\true\punycode\src)
iacaml/autolink in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\iamcal\lib_autolink)
sabre/vobject in (D:\APPLICATIFS\GLPI\WEBTEST\vendor\sabre\vobject\lib)

LDAP directories
 
Server: 'coucou.intra', Port: '389', BaseDN: 'DC=coucou,DC=intra', Connection filter:
		'(&(objectClass=user)(employeeNumber=*)(!(userAccountControl:1.2.840.113556.1.4.803:=2)))', RootDN: 'infglpi@coucou.intra',
		Use TLS: none

SQL replicas
 
Not active

Notifications
 
Way of sending emails: SMTP (anonymous@smtp)

Mails receivers
 
Name: 'infglpi@coucou.intra' Active: No
	Server: '{pop:110/pop/novalidate-cert/notls}' Login: 'COUCOU\infglpi' Password: No

Plugins list
 
	fusioninventory      Name: FusionInventory                Version: 9.1+1.0    State: Enabled

`

### PR DESCRIPTION
**Pull request by AnaelMobilia**
Fix #1476 : if we attach an already existing document to a ticket, the attachment date is the file upload date & the user logged is the uploader

** Requiered #1560 to be merged first **

### CODE REVIEW

**Comment by AnaelMobilia** [comment]
@orthagh As I known, we haven't the information of the people who attached a document to the ticket.

If we don't do anything, sender should appears as "Demandeur". (which isn't true)
Else we can arbitrary set it to the people who uploaded the document (which is false too).

An opinion about that ? (la peste ou le choléra... ?)

**Comment by orthagh** [comment]
No, it's not my point.

On migration (https://github.com/glpi-project/glpi/pull/1563/files#diff-1454c423049813362d0918fb1f2dc346R216) we should copy glpi_documents.users_id into glpi_documents_items.users_id.

You are currently adding a new field for existing data, and you can fill it with parent data.
We have some line where user can be ambiguous but no user is worse.

With your modification, we have a better relation between the user and uploaded documents. So it's good, but fill your new field.

File: install/mysql/glpi-9.2-empty.sql [hunk 1]
```diff
@@ -2086,6 +2086,7 @@ CREATE TABLE `glpi_documents_items` (
   `entities_id` int(11) NOT NULL DEFAULT '0',
   `is_recursive` tinyint(1) NOT NULL DEFAULT '0',
   `date_mod` datetime DEFAULT NULL,
+  `users_id` int(11) DEFAULT '0',
   PRIMARY KEY (`id`),
   UNIQUE KEY `unicity` (`documents_id`,`itemtype`,`items_id`),
   KEY `item` (`itemtype`,`items_id`,`entities_id`,`is_recursive`)
```
Comment (orthagh) [inline install/mysql/glpi-9.2-empty.sql]: missing key also.

File: install/update_91_92.php [hunk 2]
```diff
@@ -213,6 +213,9 @@ function update91to92() {
    // add kb category to task categories
    $migration->addField("glpi_taskcategories", "knowbaseitemcategories_id", "integer");
 
+   // #1476 - Add users_id on glpi_documents_items
+   $migration->addField("glpi_documents_items", "users_id", "integer", ['null' => "NULL"]);
+
```
Comment (orthagh) [inline install/update_91_92.php]: add a key (index) for your field

File: front/document_item.form.php [hunk 3]
```diff
@@ -44,6 +44,11 @@
 
 if (isset($_POST["add"])) {
    $document_item->check(-1, CREATE, $_POST);
+
+   // #1476 - Inject ID of the actual user to known who attach an already existing document
+   // to another item
+   $_POST['users_id'] = $_SESSION["glpiID"];
```
Comment (orthagh) [inline front/document_item.form.php]: last review, other part is good for me.
We prefer to use Class::PrepareInputForAdd instead populating data in front/ files.
So, API queries will have also the good users_id filled

Comment (AnaelMobilia) [inline front/document_item.form.php]: @orthagh I made the request changes.
Could you, please, check that it's correct ? (fisrt time I'm changing a prepareInputForAdd() method ;-))

Thanks !

Comment (orthagh) [inline front/document_item.form.php]: to be perfect, you may use this code:
```php
      if (!isset($input['users_id'])) {
         $input['users_id'] = Session::getLoginUserID();
      }
```

Comment (orthagh) [inline front/document_item.form.php]: see ProjectTask::prepareInputForAdd for example

**Comment by orthagh** [comment]
Great, thank you for your patience.

File: front/document_item.form.php [hunk 4]
```diff
@@ -44,6 +44,7 @@
 
 if (isset($_POST["add"])) {
    $document_item->check(-1, CREATE, $_POST);
+
```
Comment (orthagh) [inline front/document_item.form.php]: oh remove this ^^

**Comment by trasher** [comment]
This PR has been opened against master branch; but it should reach the 9.1/bugfixes, no?

**Comment by trasher** [comment]
Answering to myself : no, since this require a change that is present on master only

**Comment by orthagh** [comment]
Exactly :)

**Pull request merged**